compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenInterfaces.td

```diff
@@ -116,4 +116,44 @@ def IREECodegen_LoweringConfigAttrInterface :
   ];
 }
 
+def IREECodegen_LayoutAttrInterface :
+  AttrInterface<"LayoutAttrInterface"> {
+  let cppNamespace = "::mlir::iree_compiler::IREE::Codegen";
+  let description = [{
+    An interface that collects a set of methods for encoding materialization.
+  }];
+
+  let methods = [
+    InterfaceMethod<
+      /*desc=*/[{
+        Returns the layout of materialized encoding for a tensor type.
+      }],
+      /*retTy=*/"::mlir::iree_compiler::IREE::Codegen::MaterializeEncodingInfo",
+      /*methodName=*/"getEncodingInfo",
+      /*args=*/(ins "::mlir::RankedTensorType":$type),
+      /*methodBody=*/"",
+      /*defaultImplementation=*/[{
+        assert(false && "unimplemented interface method");
+        return MaterializeEncodingInfo{};
+      }]
+    >,
+    InterfaceMethod<
+      /*desc=*/[{
+        Returns the layout of materialized encoding for a tensor type.
+      }],
+      /*retTy=*/"::llvm::SmallVector<::mlir::Operation *>",
+      /*methodName=*/"lowerComputeOp",
```

Comment:
I'm thinking that we could have more interface methods for different kinds of operations. We can follow what we have in [materialization patterns](https://github.com/iree-org/iree/blob/746ad1efa3580ed53705e1246a29c749bf1e545b/compiler/src/iree/compiler/Codegen/Common/MaterializeEncodingIntoPackUnPack.cpp#L919-L936) for now:
- lowerContractionOp
- lowerSetEncodingOp
- lowerUnsetEncodingOp
- lowerShapeIndependentOp

What do you think?

Reply:
The current revision only uses it for contraction. We can rename the method to lowerContractionOp. As long as we have other needs when I implement the attribute for other backends, we can add the rest of methods.

Reply:
The current revision only uses it for contraction. We can rename the method to lowerContractionOp.  We can add the rest of methods, when we have needs for the backends.

Reply:
Use an interface method when a virtual method on the base class is needed, that is, when it is going to have overrides in more than one child class, and one wants to dispatch to them from a base class object. In that case, virtual methods are preferable to manual `if (dyn_cast<...>(...))` chains.

This becomes redundant in situations where the object, that the method would be called on, has already been downcasted to a child class.

Outside of the above specific case where an interfaceMethod is wanted, default to preferring non-member functions defined in C++, not TD.

Reply:
I see, thanks! Let me rename it to `lowerOp` and make it return a `FailureOr<Operation*>`.

Reply:
Actually, I think returning `Operation *` is good enough.